### ISSUE TITLE
[Solved] Why should I have to import models???

### ISSUE THREAD

**Issue by rconstantine** [vuex-orm/vuex-orm#462]
In your examples of both retrieving and inserting/updating, you import the model definitions. If we've already added them to the Vuex store, why do we have to import them into our vue files???

Shouldn't I be able to use this.$store instead? Or what am I missing?

My models are actually in the my store folder. I have:

root/store/database-taskSchedule

I have a Task object which holds several other objects. I assume that if I'm not accessing the children directly, then I only need to import the Task into my vue file. Is that right?

I did look at this: https://vuex-orm.github.io/vuex-orm/guide/digging-deeper/vuex-module.html but I am just not getting something. Do you have any complete working programs? I think it would be easier to understand a complete example.

**Comment by kiaking** [comment]
Actually you may call Vuex actions if you want to! Like `this.$store.dispatch('entities/users/inert', { data: {...} })`. This is what actually the model is calling behind the seen.

Importing class file is another thing though. It is like importing helper functions. Nothing more.

**Comment by rconstantine** [comment]
I see. I found this: https://github.com/vuex-orm/vuex-orm-examples

but I think it is too simple. I have this parent model:
```javascript
import { Model } from '@vuex-orm/core'
import RegInfo from './regInfo'
import Settings from './settings'
import Principal from './principal'
import { Triggery } from './triggerHierarchy'
import { Action } from './actionHierarchy'

export default class Task extends Model {
  // This is the name used as module name of the Vuex Store.
  static entity = 'task'

  // List of all fields (schema) of the post model. `this.attr` is used
  // for the generic field type. The argument is the default value.
  static fields () {
    return {
      id: this.increment(),
      name: this.string(''),
      location: this.string(''),
      facility: this.string(''),
      reginfo: this.hasOne(RegInfo, 'task_id'), // do i need to reverse this and put the hasOneBy on the other end?
      settings: this.hasOne(Settings, 'task_id'),
      principal: this.hasOne(Principal, 'task_id'),
      triggers: this.hasMany(Triggery, 'task_id'),
      actions: this.hasMany(Action, 'task_id')
    }
  }
}
```
I have imported the Task model into my Vue file. I process incoming data from a data source so that it matches the above model and its children, with the has many items as objects in an array. When I Insert the object tree, only the top level items get inserted (i.e. name, location, facility). I must be doing the insert wrong, or structuring my data wrong. The full data tree is too long to post, and it isn't formatted nicely (I'll try to format what I paste), but this is the first little bit (stringified) (note I tried reginfo and settings a bit differently, but neither worked. Also note that in object form, element names aren't quoted and values are single-quoted.):

```
[
  {
    "id":0,
    "name":"abx3_3days CVLB",
    "location":"SutterNow\\CVLB",
    "facility":"CVLB",
    "reginfo":[
      {
        "author":"SUTTER-CHS\\adtadmin",
        "date":"2014-05-23T11:37:59.4150079"
      }
    ],
    "settings: {
      "enabled":true,
      "allowHardTerminate":true,
      "allowStartOnDemand":true,
      "disallowStartIfOnBatteries":true,
      "executionTimeLimit":"PT1H",
      "hidden":false,
      "idleSettingsRestart":false,
      "idleSettingsStop":true,
      "multipleInstancesPolicy":"IgnoreNew",
      "priority":7,
      "runOnlyIfIdle":false,
      "runOnlyIfNetworkAvailable":false,
      "startWhenAvailable":true,
      "stopIfGoingOnBatteries":true,
      "wakeToRun":false
    },
    "runAsUser":{
      "userID":"SUTTER-CHS\\adtadmin",
      "runLevel":"HighestAvailable",
      "logonType":"Password"
    },
    "actions":[
      {
        "id":0,
        "type":"EXEC",
        "arguments":"rn-abx3_3days fa-cvlb fn-d:\\sutterNow_pdf\\cvlb\\abx3_3days.pdf",
        "command":"d:\\SutterNow_Reports\\SutterNow_Reports.exe",
        "workingDirectory":"d:\\SutterNow_Reports"
      }
    ],
    "triggers":[
      {
        "id":0,
        "type":"CALENDARY",
        "enabled":true,
        "randomDelay":"PT5M",
        "scheduleByDay":{
          "daysInterval":1
         },
        "startBoundary":"2014-05-23T07:01:00"
      }
    ]
} ...
```

**Comment by rconstantine** [comment]
BTW, the above object (named newJobs) is inserted like this:
```javascript
Task.insert({ data: newJobs })
```

**Comment by kiaking** [comment]
That is strange. It should insert all relationships. Is there by any chance you are looking at Vuex Dev Tool to check if the data is being inserted? In that case, children will be stored separately so it will not show up under tasks entities. It will show up inside actions and settings and such.

**Comment by rconstantine** [comment]
To check insertions, I was calling this: 
      console.log('tasks', Task.all())

Shouldn't that show the whole tree?

**Comment by rconstantine** [comment]
I'm adding an import of the RegInfo and a call to 
      console.log('regInfo', RegInfo.all())

We'll see if the data is actually getting in there or not, I suppose.

What information could I post here that would help figure this out?

**Comment by rconstantine** [comment]
OK. It looks like the data has landed in RegInfo. Why is it showing as null when part of the Task? Above, in the third post, is the Task definition. Here is the RegInfo file:
```javascript
import { Model } from '@vuex-orm/core'

export default class RegInfo extends Model {
  // This is the name used as module name of the Vuex Store.
  static entity = 'reginfo'

  // List of all fields (schema) of the post model. `this.attr` is used
  // for the generic field type. The argument is the default value.
  static fields () {
    return {
      id: this.increment(),
      task_id: this.attr(null),
      author: this.string(''),
      date: this.string('')
    }
  }
}
```
As you can see, I've given it an entity name of 'reginfo', all lowercase. I assume this is what I use in Task, like this: ```reginfo: this.hasOne(RegInfo, 'task_id'),```

Note that tasks and reginfos both have an extra property that aren't part of my model, namely $id whereas I defined just id. I assume this is normal. $id matches id.

Also note that when I build the object containing all of the data, I do NOT populate the id fields of the children because they are increment fields in the definitions.

As you can see above, reginfo has a reference to its parent task via task_id. This appears to be populated correctly.

**Question**: in my Task definition I have a 'hasOne'. Do I also need to add a 'hasOneBy' on the RegInfo?

**Comment by rconstantine** [comment]
I have managed to console.log all of the other parts of the Task object. They all are in the ORM DB (reginfo, principal, settings, actions, triggers).

Therefore, the only issue I'm having is for them to show up in the Task itself, per the above.

**Comment by rconstantine** [comment]
Solved. Your hint to see if the DB was being populated, even though I couldn't see the children in the Task, led me to know they were indeed getting in the DB. I then turned to rereading the section in the docs about retrieving. It seems my whole issue was not using .with('reginfo'). I'm now chaining several .with clauses to have it all show and it's working fine.

Thanks!

One thing that is true each time I learn a new module, is that I can't remember everything in the documentation the first time I read it all. So forgive me for not remembering the with clause until now.

Again, thanks. I believe I can now get back to just using it in my application.
